File: src/util/async/FxTimer.java


package util.async;

import javafx.animation.Animation;
import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.util.Duration;
import org.reactfx.util.Timer;

/**
* Provides factory methods for timers that are manipulated from and execute
* their action on the JavaFX application thread.
* <p>
* @author Tomas Mikula
*/
public class FxTimer implements Timer {
    
    private final Timeline timeline;
    private final Runnable action;
    
    private Duration timeout;
    private long seq = 0;
    
    /**
    * Creates a (stopped) timer that executes the given action specified number
    * of times with a delay period.
    * @param delay Time to wait before each execution. The first execution is 
    * already delayed.
    * @param action action to execute
    * @param cycles denotes number of executions. Use 1 for single execution, n
    * for n executions and {@link Animation.INDEFINITE} (-1) for infinite amount.
    */
    public FxTimer(Duration delay, int cycles, Runnable action) {
        this.timeout = Duration.millis(delay.toMillis());
        this.timeline = new Timeline();
        this.action = action;

        timeline.setCycleCount(cycles);
    }
    /**
    * Equivalent to {@code new FxTimer(Duration.millis(delay), action, cycles);}
    */
    public FxTimer(double delay, int cycles, Runnable action) {
        this(Duration.millis(delay), cycles, action);
    }

    @Override
    public void restart() {
        restart(timeout);
    }
    
    /**
     * Equivalent to calling {@link #setTimeout()} and {@link #restart()}
     * subsequently.
     * @param timeout 
     */
    public void restart(Duration timeout) {
        stop();
        long expected = seq;
        this.timeout = timeout;
        timeline.getKeyFrames().setAll(new KeyFrame(timeout, ae -> {
            if(seq == expected) {
                action.run();
            }
        }));
        timeline.play();
    }

    public void restart(double timeoutInMillis) {
        restart(Duration.millis(timeoutInMillis));
    }
    
    @Override
    public void stop() {
        timeline.stop();
        seq++;
    }
    
    public boolean isRunning() {
        return timeline.getCurrentRate()!=0;
    }
    
    /**
     * Sets the delay for the task. Takes effect only if set before the task
     * execution is planned. It will not affect currently running cycle. It will
     * affect every subsequent cycle. Therefore, it is pointless to run this
     * method if this timer is non-periodic.
     * @param timeout 
     */
    public void setTimeout(Duration timeout) {
        this.timeout = timeout;
    }
    
    public void setTimeout(double timeautInMillis) {
        this.timeout = Duration.millis(timeautInMillis);
    }
    
    /**
     * If timer running, executes {@link #restart(javafx.util.Duration)}, else
     * executes {@link #setTimeout(javafx.util.Duration)}
     * <p>
     * Basically same as {@link #restart(javafx.util.Duration)}, but restarts
     * only if needed (when running).
     * 
     * @param timeout 
     */
    public void setTimeoutAndRestart(Duration timeout) {
        if (isRunning()) restart(timeout);
        else setTimeout(timeout);
    }
    
    public void setTimeoutAndRestart(double timeautInMillis) {
        if (isRunning()) restart(Duration.millis(timeautInMillis));
        else setTimeout(Duration.millis(timeautInMillis));
    }
}

File: src/util/main/sp/it/util/async/executor/FxTimer.kt
/*
 * Implementation based on ReactFX by Tomas Mikula, https://github.com/TomasMikula/ReactFX
 */

package sp.it.util.async.executor

import javafx.animation.KeyFrame
import javafx.animation.Timeline
import javafx.util.Duration
import sp.it.util.collections.setToOne
import sp.it.util.functional.invoke
import sp.it.util.units.millis

/**
 * Lightweight timer constrained to JavaFX application thread.
 * Allows running an action with delay or periodically as well as restarting or querying state.
 *
 * Not thread safe! Must only be used on FxApplication thread.
 */
class FxTimer {

   /**
    * Sets the delay for the task. Takes effect only if set before the task
    * execution is planned. It will not affect currently running cycle. It will
    * affect every subsequent cycle. Therefore, it is pointless to run this
    * method if this timer is non-periodic.
    */
   var period: Duration
   private val action: Runnable
   private val timeline: Timeline
   private var seq: Long = 0

   /** Returns true if not stopped or paused */
   val isRunning: Boolean
      get() = timeline.currentRate!=0.0

   private constructor(delay: Duration, cycles: Int, action: Runnable) {
      this.action = action
      this.period = delay
      this.timeline = Timeline()
      this.timeline.cycleCount = cycles
   }

   /** @param run if true, [start] will be called, otherwise calls [stop]. */
   fun setRunning(run: Boolean) {
      if (run) start()
      else stop()
   }

   /** Equivalent to setting [period] and then calling [start]. */
   @JvmOverloads
   fun start(period: Duration = this.period) {
      stop()
      val expected = seq
      this.period = period

      if (period.toMillis()==0.0) {
         runNow()
      } else {
         timeline.keyFrames setToOne KeyFrame(period, {
            if (seq==expected)
               runNow()
         })
         timeline.play()
      }
   }

   fun runNow() = action()

   fun pause() = timeline.pause()

   fun unpause() = timeline.play()

   fun stop() {
      timeline.stop()
      seq++
   }

   /**
    * If timer running, executes [start], otherwise sets the period.
    *
    * Essentially same as [start], but restarts only when already running
    */
   fun setTimeoutAndRestart(timeout: Duration) {
      if (isRunning)
         start(timeout)
      else
         period = timeout
   }

   fun setTimeoutAndRestart(timeoutMs: Double) {
      if (isRunning) start(timeoutMs.millis)
      else period = timeoutMs.millis
   }

   companion object {
      /**
       * Creates a (stopped) timer that executes the given action a specified number of times with a delay period.
       *
       * @param delay time to wait before each execution. The first execution is already delayed.
       * @param cycles denotes number of executions, use -1 for infinite executions
       * @param action action to execute
       */
      @JvmStatic
      fun fxTimer(delay: Duration, cycles: Int, action: () -> Unit) = FxTimer(delay, cycles, action)
   }

}